rs/rosetta-api/icp/tests/system_tests/test_cases/voting.rs

```diff
@@ -0,0 +1,333 @@
+        );
+
+        // Create neurons
+        let neuron_ids = create_neurons(&env, INITIAL_BALANCE / 10, DISSOLVE_DELAY_6_MONTHS).await;
+
+        // Ensure no proposals exist initially
+        assert!(get_pending_proposals(&env).await.unwrap().is_empty());
+
+        // Submit multiple proposals
+        for i in 0..3 {
+            governance_client
+                .submit_proposal(
+                    TEST_IDENTITY.sender().unwrap(),
+                    neuron_ids[0].into(),
+                    &format!("dummy title {}", i),
+                    &format!("test {}", i),
+                    &format!("dummy text {}", i),
+                )
+                .await;
+        }
+
+        // Ensure all proposals are pending and have the expected details
+        let all_proposals = get_pending_proposals(&env).await.unwrap();
+        let all_proposals_from_governance = governance_client.get_pending_proposals().await;
+
+        assert_eq!(all_proposals.len(), 3);
+
+        let expected_proposals = vec![
+            Proposal {
+                title: Some("dummy title 0".to_string()),
+                summary: "test 0".to_string(),
+                action: Some(Action::Motion(Motion {
+                    motion_text: "dummy text 0".to_string(),
+                })),
+                url: "".to_string(),
+            },
+            Proposal {
+                title: Some("dummy title 1".to_string()),
+                summary: "test 1".to_string(),
+                action: Some(Action::Motion(Motion {
+                    motion_text: "dummy text 1".to_string(),
+                })),
+                url: "".to_string(),
+            },
+            Proposal {
+                title: Some("dummy title 2".to_string()),
+                summary: "test 2".to_string(),
+                action: Some(Action::Motion(Motion {
+                    motion_text: "dummy text 2".to_string(),
+                })),
+                url: "".to_string(),
+            },
+        ];
+
+        // Verify all proposals are pending and have the expected details both from
+        // Rosetta and the governance canister directly.
+        assert_eq!(all_proposals, expected_proposals);
+        for (i, proposal) in all_proposals_from_governance.iter().enumerate() {
+            assert_eq!(proposal.proposal.clone().unwrap(), expected_proposals[i]);
+        }
+
+        let proposal_ids = all_proposals_from_governance
+            .iter()
+            .map(|proposal| proposal.id.unwrap().id)
+            .collect::<Vec<u64>>();
+
+        // Vote on the first proposal and verify the YES vote was correctly registered
+        register_vote(&env, proposal_ids[0], NEURON_INDEX[1], Vote::Yes)
+            .await
+            .unwrap();
+        let pending_proposals = governance_client.get_pending_proposals().await;
+        let pending_proposals_from_governance = governance_client.get_pending_proposals().await;
+        assert_eq!(pending_proposals.len(), 3);
```

Comment:
I ran the new test multiple times, and [it failed this assertion in 2/100 cases](https://dash.ln1-idx1.dfinity.network/invocation/72e5b412-9347-484c-a763-854f2d1c9f51). I didn't look into it in too much detail yet, but since this is a "live" pocket IC instance, the test is not deterministic, and this can lead to failures (especially on CI) e.g., if there are not enough resources. Polling for the number of pending proposals to become 3 (with a timeout) could be an option here, but maybe there are better solutions?

Reply:
The issue was with the governance client closing proposals as soon as we had 2 out of 4 votes in some cases. I wasn't able to figure out how to control this source of non-determinism in the test so, since cheking for that particular behavior is outside the scope of the test, I just removed checking it. Instead, we now only check for the actual votes in the proposals, which is actually within the scope of this test. The test now passes 100/100.